lib/src/new-select/NewSelect.jsx

```diff
@@ -0,0 +1,699 @@
+
+    const [isOpen, changeIsOpen] = useState(false);
+    const [isActiveOption, changeIsActiveOption] = useState(false);
+    const [filteredOptions, setFilteredOptions] = useState([]);
+    const [visualFocusedSuggIndex, changeVisualFocusedSuggIndex] = useState(-1);
+
+    const selectContainerRef = useRef(null);
+    const selectInputRef = useRef(null);
+
+    const colorsTheme = useTheme();
+    const backgroundType = useContext(BackgroundColorContext);
+
+    const notOptionalCheck = (value) => value === "" && !optional;
+
+    const canBeOpenOptions = () => !disabled && options && options.length; // === options?.length ???
+    const groupsHaveOptions = () =>
+      options[0].options ? filteredOptions.some((option) => option.options?.length > 0) : true;
+    const openOptions = () => {
+      if (canBeOpenOptions()) {
+        searchable && changeIsBackgroundValue(true);
+        changeIsOpen(true);
+      }
+    };
+    const closeOptions = () => {
+      searchable && changeIsBackgroundValue(false);
+      changeIsOpen(false);
+      changeVisualFocusedSuggIndex(-1);
+    };
+
+    const handleSelectChangeValue = (newValue) => {
+      value ?? setInnerValue(newValue);
+
+      if (notOptionalCheck(newValue)) onChange?.({ value: newValue, error: getNotOptionalErrorMessage() });
+      else onChange?.({ value: newValue, error: null });
+    };
+    const handleSelectOnClick = () => {
+      isOpen ? closeOptions() : openOptions();
+      searchable && selectInputRef.current.focus();
+    };
+    const handleSelectOnFocus = () => {
+      searchable && selectInputRef.current.focus();
+    };
+    const handleSelectOnBlur = (event) => {
+      // focus leaves container (outside, not to childs)
+      if (!event.currentTarget.contains(event.relatedTarget)) {
+        setSearchValue("");
+        closeOptions();
+
+        if (notOptionalCheck(value ?? innerValue))
+          onBlur?.({ value: value ?? innerValue, error: getNotOptionalErrorMessage() });
+        else onBlur?.({ value: value ?? innerValue, error: null });
+      }
+    };
+
+    const handleSearchIOnChange = (event) => {
+      openOptions();
+      setSearchValue(event.target.value);
+    };
+
+    const handleClearActionOnClick = (event) => {
+      event.stopPropagation(); // not bubbling, clear not executes select event onClick
+      setSearchValue("");
+    };
+
+    const getSingleSelectedOptionLabel = () => {
+      const val = value ?? innerValue;
+      let selectedOptionLabel;
+
+      if (options && options.length) {
+        options.forEach((option) => {
+          if (option.options) {
+            option.options.forEach((singleOption) => {
+              if (singleOption.value === val) selectedOptionLabel = singleOption.label;
+            });
+          } else if (option.value === val) selectedOptionLabel = option.label;
+        });
+      }
+
+      return selectedOptionLabel;
+    };
+
```

Comment:
I think it could be interesting to encapsulate this function in a useMemo. Now this function is being executed always (when the select is opened, when we make hover of an option, when we select an option...). Something like this `const selectedLabel = useMemo(() => getSingleSelectedOptionLabel(), [value ?? innerValue]);` (and use this `selectedLabel` in the render instead of calling the function) will make that this is only called when the value changes.

Reply:
Great idea, I will implement it, thanks!